### ISSUE TITLE
[web]: PageView.builder has a different behavior on Web and mobile

### ISSUE THREAD

**Issue by Neamar** [flutter/flutter#54052]
tl;dr: widget's dispose() are called at different times for the discarded pages.

## Steps to Reproduce

1. Run `flutter create bug`.
2. Replace main file with:

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: PageView.builder(
          itemBuilder: (context, position) {
            return SamplePage(ValueKey(position), position);
          },
          itemCount: 3,
        ),
      ),
    );
  }
}

class SamplePage extends StatefulWidget {
  final int pageIndex;

  SamplePage(Key key, this.pageIndex) : super(key: key);

  @override
  _SamplePageState createState() => _SamplePageState(pageIndex);
}

class _SamplePageState extends State<SamplePage> {
  final int pageIndex;
  int count = 0;

  _SamplePageState(this.pageIndex);

  @override
  Widget build(BuildContext context) {
    return RaisedButton(
      child: Text("Page " + pageIndex.toString() + ".\nClicked " + count.toString() + " times"),
      onPressed: () {
        setState(() {
          count++;
        });
      },
    );
  }
}
```
3. Tap the widget on the middle of the screen. This will display "Page 0. Clicked 1 times"
4. Swipe right to display next page. Tap the widget, this will display "Page1. Clicked 1 times".
5. Swipe back to page 1.

**Expected results:** widget for page 0 is reset to its initial state. This is what happens on mobile: as soon as an element disappears from view, its `dispose()` function is called.

**Actual results:** on web, if you scroll back the previous widget is still displayed. You can add multiple pages, and moving back to them will display them with the states they had the last time.

I don't think either behaviors are bad, but I'm expecting the same behavior for both.
I looked into various attributes (and the `keepPage` parameter from PageController) but it doesn't look like this is expected behavior.

I'm new to Flutter, and I hope I didn't miss anything obvious. If so, feel free to close, sorry!

<details>
  <summary>Logs</summary>
`flutter analyze`:

```
No issues found! (ran in 4.4s)
```

```
[✓] Flutter (Channel beta, v1.15.17, on Linux, locale en_US.UTF-8)
    • Flutter version 1.15.17 at /home/neamar/Apps/flutter
    • Framework revision 2294d75bfa (4 weeks ago), 2020-03-07 00:28:38 +0900
    • Engine revision 5aff311948
    • Dart version 2.8.0 (build 2.8.0-dev.12.0 9983424a3c)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /home/neamar/Android/Sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 28.0.3
    • Java binary at: /home/neamar/Apps/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Android Studio (version 3.6)
    • Android Studio at /home/neamar/Apps/android-studio
    • Flutter plugin version 45.0.1
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[✓] Connected device (3 available)
    • Pixel 2    • HT82D1A02300 • android-arm64  • Android 10 (API 29)
    • Chrome     • chrome       • web-javascript • Google Chrome 80.0.3987.149
    • Web Server • web-server   • web-javascript • Flutter Tools

• No issues found!

```

</details>

I had the same issue running on `master`.

**Comment by TahaTesser** [comment]
Hi @Neamar 
Tested your code sample, it seems to be work as expected  
checkout this [dartpad](https://dartpad.dev/8547399d9754a4df4e8b47532f567791)

Page 0 resets to 0 times after swiping back 
let me know if i missed something


```
[✓] Flutter (Channel dev, v1.18.0, on Mac OS X 10.15.4 19E266, locale en-GB)
    • Flutter version 1.18.0 at /Users/taha/Code/flutter_dev
    • Framework revision 14f3a36aba (2 days ago), 2020-04-03 17:06:01 -0700
    • Engine revision 09bc1fc45e
    • Dart version 2.8.0 (build 2.8.0-dev.20.0 e736495eb7)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/taha/Code/sdk
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = /Users/taha/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.4, Build version 11E146
    • CocoaPods version 1.9.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 45.0.1
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build
      1.8.0_212-release-1586-b4-5784211)

[✓] VS Code (version 1.43.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.9.0

[✓] Connected device (4 available)
    • Taha’s iPhone • 00008020-001059882212002E • ios            • iOS 13.4
    • macOS         • macOS                     • darwin-x64     • Mac OS X
      10.15.4 19E266
    • Chrome        • chrome                    • web-javascript • Google Chrome
      80.0.3987.149
    • Web Server    • web-server                • web-javascript • Flutter Tools

• No issues found!
```

Thank you

**Comment by Neamar** [comment]
Hi @TahaTesser, thanks for taking the time to test! Looks like I slightly messed up in the code I send, my bad.

Version that I should have shared is below,

It's a very simple thing, each page has a Switch, when you click the switch you get a green banner "You won". If you scroll next from there, after displaying the "You won state" (stored in `gameState`), the widget will never be disposed.

Steps to reproduce, on web:
1. Click the toggle. Get green banner
2. Move to next screen. Notice there is no `console.log()` from the `dispose()` function
3. Move to third screen. This time, you'll see a call to `dispose()`, for the second screen.
4. Move back to first screen. First screen will still be in the same state. All other widgets will be disposed.

Steps to reproduce, on mobile:
1. Click the toggle. Get green banner
2. Move to next screen. Notice page is properly disposed, unlike in web.


I spent quite some time boiling it down to something relatively simple, and I'm now convinced that I'm missing something fundamental on lifecycle management for Flutter widget. Still, behavior is different for web and mobile, which is probably not expected.

Updated dartpad: https://dartpad.dev/f81c8888fa9eaaa9825b09106d4c9fe6

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: PageView.builder(
          itemBuilder: (context, position) {
            Key key = ValueKey(position);
            return Level(key);
          },
          itemCount: 3,
        ),
      ),
    );
  }
}

class Level extends StatefulWidget {
  Level(Key key) : super(key: key);

  @override
  _LevelState createState() => _LevelState();
}

class _LevelState extends State<Level> {
  static const MaterialColor COLOR_GAME = Colors.deepPurple;
  static const MaterialColor COLOR_SUCCESS = Colors.green;

  static const STATE_PLAYING = 'playing';
  static const STATE_WON = 'won';

  String gameState;

  _LevelState() {
    gameState = STATE_PLAYING;
  }

  @override
  void dispose() {
    print("Disposing of level, state = " + gameState);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    MaterialColor headerColor = COLOR_GAME;
    String textToDisplay = "Click the toggle to win";
    if (gameState == STATE_WON) {
      headerColor = COLOR_SUCCESS;
      textToDisplay = "You won!";
    }

    return Column(children: <Widget>[
      Material(
        color: headerColor[300],
        child: InkWell(
          child: Padding(
            padding: const EdgeInsets.all(8.0),
            child: Text(textToDisplay),
          ),
        ),
      ),
      Expanded(
        child: SwitchListTile(
          title: Text("Make me win"),
          value: gameState != STATE_PLAYING,
          onChanged: gameState != STATE_PLAYING
              ? null
              : (bool value) {
            setState(() {
              gameState = STATE_WON;
            });
          },
        ),
      )
    ]);
  }
}
```

**Comment by TahaTesser** [comment]
<details> 
<summary>flutter doctor -v</summary> 

```
[✓] Flutter (Channel dev, 1.21.0-7.0.pre, on Mac OS X 10.15.6 19G73, locale
    en-GB)
    • Flutter version 1.21.0-7.0.pre at /Users/tahatesser/Code/flutter_dev
    • Framework revision 5a6dfa35ca (9 days ago), 2020-08-03 10:33:07 -0700
    • Engine revision 083282e33b
    • Dart version 2.10.0 (build 2.10.0-4.0.dev 365525432a)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.1
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 48.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.47.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.13.2

[✓] Connected device (6 available)
    • SM M305F (mobile)      • 32003c30dc19668f                     •
      android-arm64  • Android 10 (API 29)
    • Taha’s iPhone (mobile) • 00008020-001059882212002E            • ios
      • iOS 13.6
    • iPhone 11 (mobile)     • A63D6F51-59F3-4A75-961F-3769709F3A92 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-13-6 (simulator)
    • macOS (desktop)        • macos                                • darwin-x64
      • Mac OS X 10.15.6 19G73
    • Web Server (web)       • web-server                           •
      web-javascript • Flutter Tools
    • Chrome (web)           • chrome                               •
      web-javascript • Google Chrome 84.0.4147.125

• No issues found!

```

</details>

**Comment by exaby73** [comment]
I cannot reproduce this issue on the latest master `3.5.0-10.0.pre.17`. The state is disposed of on web and mobile. Closing this issue as fixed for the same reason.
